docs/cow-protocol/reference/contracts/periphery/composable_cow.md

```diff
@@ -0,0 +1,520 @@
+|---|---|
+| `order` | Proposed discrete order |
+| `ctx` | [Execution context](#execution-context) |
+| `params` | [`ConditionalOrderParams`](#conditionalorderparams-struct) |
+| `offchainInput` | Conditional order type-specific data **NOT** known at time of creation for a **specific** discrete order (or zero-length bytes if not applicable) |
+
+### Guarantees and Invariants
+
+- CoW Protocol's settlement contract enforces single-use orders, ie. **NO** `GPv2Order` can be filled more than once
+- For merkle trees, `H(ConditionalOrderParams)` **MUST** be a member of the merkle tree `roots[owner]`
+- For single orders, `singleOrders[owner][H(ConditionalOrderParams)] == true`
+
+## Data Types and Storage
+
+### `ConditionalOrderParams`
+
+A conditional order is defined by the following data:
+
+```solidity
+struct ConditionalOrderParams {
+    IConditionalOrder handler;
+    bytes32 salt;
+    bytes staticData;
+}
+```
+
+| **Field** | **Description** |
+|---|---|
+| `handler` | The contract implementing the conditional order logic |
+| `salt` | Allows for multiple conditional orders of the same type and data |
+| `staticData` | Data available to **ALL** *discrete* orders created by the conditional order |
+
+:::note
+
+All of the above fields are verified by `ComposableCoW` to be valid, prior to calling the `verify` method on the handler (`IConditionalOrder`).
+
+:::
+
+:::tip
+
+When used with Merkle Trees and a cryptographically-secure random `salt`, the conditional order is effectively private (until a discrete order cut from this conditional order is broadcast to the CoW Protocol API).
+
+:::
+
+:::caution
+
+* `H(ConditionalOrderParams)` **MUST** be unique
+* Not setting `salt` to a cryptographically-secure random value **MAY** result in leaking information or hash collisions
+
+:::
+
+### `PayloadStruct`
+
+This is the data passed to `ComposableCoW` via the `payload` parameter of `isValidSafeSignature`:
+
+```solidity
+struct PayloadStruct {
+    bytes32[] proof;
+    IConditionalOrder.ConditionalOrderParams params;
+    bytes offchainInput;
+}
+```
+
+| **Field** | **Description** |
+|---|---|
+| `proof` | Merkle Tree proof (if applicable, zero length otherwise) |
+| `params` | [`ConditionalOrderParams`](#conditionalorderparams-struct) |
+| `offchainInput` | Off-chain input (if applicable, zero length otherwise) |
+
+By setting `proof` to zero-length, this indicates to `ComposableCoW` that the order is a single order, and not part of a Merkle Tree.
+
+### `Proof`
+
+```solidity
+struct Proof {
+    uint256 location;
+    bytes data;
+}
+```
+
+:::note
+
+The `Proof.location` is intentionally not made an `enum` to allow for future extensibility as other proof locations may be integrated.
+
+:::
+
+| **Field** | **Description** |
+|---|---|
+| `location` | An integer representing the location where to find the proofs |
+| `data` | `location` implementation specific data for retrieving the proofs |
+
+#### Locations
+
```

Comment:
```suggestion

Some services pick up new conditional orders automatically from on-chain events.
The proof data can be emitted onchain, but it can also be retrieved from other supported on-chain services.
The location field signals where this data can be retrieved. 

```